src/Data/Options.purs

```diff
@@ -1,94 +1,97 @@
 module Data.Options
   ( Options()
-  , Option()
-  , IsOption
-  , assoc, (:=)
-  , optionFn
-  , options
   , runOptions
+  , options
+  , Option(..)
+  , assoc, (:=)
+  , massoc, (:=?)
   , opt, key
+  , defaultToOptions
   ) where
 
 import Prelude
 
-import Data.Foreign (Foreign())
-
-import Data.Function (Fn2(), runFn2)
+import Data.Foreign (toForeign, Foreign())
 
-import Data.Maybe (Maybe(..))
+import Data.Maybe (Maybe(), maybe)
 
-import Data.Monoid (Monoid)
+import Data.Monoid (mempty, Monoid)
 
 import Data.Tuple (Tuple(..))
 
-import Unsafe.Coerce (unsafeCoerce)
+import Data.StrMap as StrMap
 
-foreign import data Options :: * -> *
+import Data.Functor.Contravariant (Contravariant)
 
-foreign import data Option :: * -> * -> *
+-- | The `Options` type represents a set of options. The type argument is a
+-- | phantom type, which is useful for ensuring that options for one particular
+-- | API are not accidentally passed to some other API.
+newtype Options opt =
+  Options (Array (Tuple String Foreign))
 
-class IsOption value where
-  assoc :: forall opt. Option opt value -> value -> Options opt
+runOptions :: forall opt. Options opt -> Array (Tuple String Foreign)
+runOptions (Options xs) = xs
 
-infixr 6 :=
-
-(:=) :: forall opt value. (IsOption value) => Option opt value -> value -> Options opt
-(:=) = assoc
+instance semigroupOptions :: Semigroup (Options opt) where
+  append (Options xs) (Options ys) = Options (xs <> ys)
 
-instance semigroupOptions :: Semigroup (Options a) where
-  append = runFn2 appendFn
+instance monoidOptions :: Monoid (Options opt) where
+  mempty = Options []
 
-instance monoidOptions :: Monoid (Options a) where
-  mempty = memptyFn
+-- | Convert an `Options` value into a JavaScript object, suitable for passing
+-- | to JavaScript APIs.
+options :: forall opt. Options opt -> Foreign
+options = toForeign <<< StrMap.fromFoldable <<< runOptions
 
-instance isOptionString :: IsOption String where
-  assoc = runFn2 isOptionPrimFn
+-- | An `Option` represents an opportunity to configure a specific attribute
+-- | of a call to some API. This normally corresponds to one specific property
+-- | of an "options" object in JavaScript APIs.
+newtype Option opt value =
+  Option
+    { key :: String
+    , toOptions :: String -> value -> Options opt
```

Comment:
Sorry! I didn't seem to get the notification on this. I don't really recall anything specifically to do with `key`, but it's been a while since I made use of `-options` so it's a little hard to remember.

I've been trying to think where I might have used it, other than tweaking things in `gulp-purescript`, I feel like there is somewhere I'm forgetting about where I wrote some stuff from scratch and needed some assistance to get it to do what I wanted, but I have no idea what it was now.